include/rocksdb/customizable.h

```diff
@@ -0,0 +1,149 @@
+  // the default implementation.  This method should be used when you need a
+  // specific derivative or implementation of a class.
+  //
+  // Intermediary caches (such as SharedCache) may wish to override this method
+  // to check for the intermediary name (SharedCache).  Classes with multiple
+  // potential names (e.g. "PosixEnv", "DefaultEnv") may also wish to override
+  // this method.
+  //
+  // @param name The name of the instance to find.
+  // Returns true if the class is an instance of the input name.
+  virtual bool IsInstanceOf(const std::string& name) const {
+    return name == Name();
+  }
+
+  // Returns the named instance of the Customizable as a T*, or nullptr if not
+  // found. This method uses IsInstanceOf to find the appropriate class instance
+  // and then casts it to the expected return type.
+  template <typename T>
+  const T* CastAs(const std::string& name) const {
+    if (IsInstanceOf(name)) {
+      return static_cast<const T*>(this);
+    } else {
+      const auto inner = static_cast<const Customizable*>(Inner());
+      if (inner != nullptr) {
+        return inner->CastAs<T>(name);
+      } else {
+        return nullptr;
+      }
+    }
+  }
+
+  template <typename T>
+  T* CastAs(const std::string& name) {
+    if (IsInstanceOf(name)) {
+      return static_cast<T*>(this);
+    } else {
+      const auto inner = static_cast<const Customizable*>(Inner());
+      if (inner != nullptr) {
+        return (const_cast<Customizable*>(inner))->CastAs<T>(name);
+      } else {
+        return nullptr;
+      }
+    }
+  }
+
+  // Checks to see if this Configurable is equivalent to other.
+  // This method assumes that the two objects are of the same class.
+  // @param config_options Controls how the options are compared.
+  // @param other The other object to compare to.
+  // @param mismatch If the objects do not match, this parameter contains
+  //      the name of the option that triggered the match failure.
+  // @param True if the objects match, false otherwise.
+  bool AreEquivalent(const ConfigOptions& config_options,
+                     const Configurable* other,
+                     std::string* name) const override;
+#ifndef ROCKSDB_LITE
+  Status GetOption(const ConfigOptions& config_options, const std::string& name,
+                   std::string* value) const override;
+
+#endif  // ROCKSDB_LITE
+ protected:
+  std::string GetOptionName(const std::string& long_name) const override;
+#ifndef ROCKSDB_LITE
+  std::string SerializeOptions(const ConfigOptions& options,
+                               const std::string& prefix) const override;
+#endif  // ROCKSDB_LITE
+ public:
+  // Helper method for configuring a new customizable object.
+  // If base_opts are set, this is the "default" options to use for the new
+  // object whereas "new_opts" are overlaid on the base options. Returns OK if
+  // the object could be successfully configured
+  static Status ConfigureNewObject(
```

Comment:
It is not very clear for me to understand the `new_opts` and `base_opts`. If `base_opts` is not set, the customizable object is configured with the `new_opts`? If `base_opts` is set, `new_opts` will be updated with the `base_opts`?

Reply:
First the object is configured using base_opts and then it is configured again using new_opts. 

Say you had something like a BlockBasedTableOptions and you want to create a new one and change only one option.  To do this generically (meaning you do not know you have a BlockBasedTableOptions but have an old class), you first serialize the existing object to a string -- this is your base_opts.  Then you configure the new object using that base_opts string.  Then you override what you wanted to change with the new_opts.

Reply:
I see, thanks for the explanation.